joomla/joomla-cms
# The One Right Session Management Configuration For Joomla! 3

### Summary of Changes

- Implements the Session MetadataManager object originally proposed in #19578
- Implements a new Session Garbage Collection plugin
    - Can optionally trigger PHP's native session garbage collection task
    - Can optionally trigger Joomla's metadata cleanup task
- Implements a new CLI script for use as a cron job for cleaning the metadata (well, actually this causes garbage collection on the session table in general, but it's a side effect of the incorrect dual purpose of the session table)

### Testing Instructions

Apply patch, if necessary discover and install the new plugin.  With the "Enable Session Garbage Collection" option enabled, during the `onAfterRespond` event Joomla will attempt to perform session garbage collection based on the odds defined in the plugin's parameters.  With the "Enable Session Metadata Cleanup" option enabled, and a non-database session handler in use, during the `onAfterRespond` event Joomla will attempt to purge stale records from the session database table representing the optional session metadata (as explained a thousand times, this is done by design in the plugin for now because the session store and our metadata handling represent two different datasets and should be treated as such).

The odds calculation is the same as that used by the C level [`php_session_gc`](https://github.com/php/php-src/blob/703e0370905bc432aa26d71803b3081851e99230/ext/session/session.c#L366-L383) function, ported to userland PHP and using the plugin's configuration values versus the PHP runtime configuration.

The two cleanup operations purposefully do not share the same probability calculation, though there is no harm in running both tasks in one request IMO there is no need to force it either.
